ptp/c4u/clock/clock.go

```diff
@@ -49,44 +59,74 @@ func (rb *RingBuffer) Write(c *ptp.ClockQuality) {
 }
 
 // Export data from the ring buffer
-func (rb *RingBuffer) Data() []*ptp.ClockQuality {
+func (rb *RingBuffer) Data() []*DataPoint {
 	return rb.data
 }
 
-func Worst(clocks []*ptp.ClockQuality) *ptp.ClockQuality {
+func Worst(points []*DataPoint, accuracyExpr string) (*ptp.ClockQuality, error) {
+	expr, err := prepareExpression(accuracyExpr)
+	if err != nil {
+		return nil, fmt.Errorf("evaluating accuracy math: %w", err)
+	}
+	phcOffsets := []float64{}
+	oscillatorOffsets := []float64{}
 	var w *ptp.ClockQuality
-	for _, c := range clocks {
+	for _, c := range points {
 		if c == nil {
 			continue
 		}
-
 		if w == nil {
 			w = &ptp.ClockQuality{}
 		}
-
-		// Higher value of accuracy means worse
-		if c.ClockAccuracy > w.ClockAccuracy {
-			w.ClockAccuracy = c.ClockAccuracy
+		phcOffsets = append(phcOffsets, float64(c.PHCOffset))
+		// if oscillator is uncalibrated, ignore the offset as it's meaningless
+		if c.OscillatorClockClass != ClockClassUncalibrated {
+			oscillatorOffsets = append(oscillatorOffsets, float64(c.OscillatorOffset))
 		}
 
+		// TODO: consider better way to select clock class
 		// Assuming higher class means worse
-		if c.ClockClass > w.ClockClass {
-			w.ClockClass = c.ClockClass
+		if c.OscillatorClockClass > w.ClockClass {
+			w.ClockClass = c.OscillatorClockClass
 		}
 	}
-	return w
+	if w == nil {
+		return nil, nil
+	}
+
+	parameters := map[string]interface{}{
+		"phcoffset":        phcOffsets,
+		"oscillatoroffset": oscillatorOffsets,
+	}
+	vRaw, err := expr.Evaluate(parameters)
```

Comment:
This is super elegant btw! Kudos!

Reply:
Although currently oscillatord offsets are always positive (100ns or 1us) vs ts2phc offset which can be any sign. Does it matter?

Reply:
It doesn't matter now, and it's easy to account for in the expression.

In the current formula we use `max(abs(mean(phcoffset)) + 3 * stddev(phcoffset), abs(mean(oscillatoroffset)))`, it's basically max between `p99(phcoffset)` and `mean(oscillatordoffset)`.